File: app/workflow/gates.py
from __future__ import annotations

import re
from typing import Any


REQUIRED_BOOLEAN_GATES = [
    "Strategic Alignment Brief",
    "Problem Quantified",
    "≥3 Options Evaluated",
    "Success Metrics Defined",
    "Leading Indicators Defined",
    "Kill Criteria Defined",
]


def _checkbox_true(prop: dict[str, Any] | None) -> bool:
    if not prop:
        return False
    return bool(prop.get("checkbox"))


def _number_present(prop: dict[str, Any] | None) -> bool:
    if not prop:
        return False
    return prop.get("number") is not None


def _select_present(prop: dict[str, Any] | None) -> bool:
    if not prop:
        return False
    sel = prop.get("select")
    return bool(sel and sel.get("name"))


def infer_governance_checks_from_text(body_text: str) -> dict[str, bool]:
    text = body_text.lower()

    def has_any(*phrases: str) -> bool:
        return any(phrase in text for phrase in phrases)

    def explicitly_no(gate_name: str) -> bool:
        return f"{gate_name.lower()}: no" in text

    option_matches = re.findall(r"\boption\s+[a-z0-9]+\b", text)
    numeric_matches = re.findall(r"\b\d[\d,\.%]*\b", text)

    return {
        "Strategic Alignment Brief": (not explicitly_no("Strategic Alignment Brief")) and has_any("strategic context", "strategic alignment", "objective supported"),
        "Problem Quantified": (not explicitly_no("Problem Quantified")) and has_any("problem framing", "quantified impact", "problem statement") and len(numeric_matches) >= 3,
        "≥3 Options Evaluated": (not explicitly_no("≥3 Options Evaluated")) and has_any("options evaluated", "chosen option") and len(set(option_matches)) >= 3,
        "Success Metrics Defined": (not explicitly_no("Success Metrics Defined")) and has_any("success metrics", "primary metric", "kpi impact"),
        "Leading Indicators Defined": (not explicitly_no("Leading Indicators Defined")) and has_any("leading indicators"),
        "Kill Criteria Defined": (not explicitly_no("Kill Criteria Defined")) and has_any("kill criteria", "we will stop or pivot"),
        "Option Trade-offs Explicit": (not explicitly_no("Option Trade-offs Explicit")) and has_any("trade-offs", "trade offs"),
        "Risk Matrix Completed": (not explicitly_no("Risk Matrix Completed")) and has_any("risk matrix") and has_any("mitigation", "probability", "impact"),
        "Financial Model Included": (not explicitly_no("Financial Model Included")) and has_any("financial model", "payback period", "revenue impact", "cost impact"),
        "Downside Modeled": (not explicitly_no("Downside Modeled")) and has_any("downside", "risk-adjusted", "sensitivity"),
        "Compliance Reviewed": (not explicitly_no("Compliance Reviewed")) and has_any("compliance review", "compliance reviewed", "legal review", "regulatory review"),
        "Decision Memo Written": (not explicitly_no("Decision Memo Written")) and has_any("executive summary", "final decision"),
        "Root Cause Done": (not explicitly_no("Root Cause Done")) and has_any("root cause"),
        "Assumptions Logged": (not explicitly_no("Assumptions Logged")) and has_any("assumptions", "confidence level"),
    }


def evaluate_required_gates(
    page_properties: dict[str, Any],
    inferred_checks: dict[str, bool] | None = None,
) -> list[str]:
    missing: list[str] = []
    if not _number_present(page_properties.get("Baseline")):
        missing.append("Baseline")
    if not _number_present(page_properties.get("Target")):
        missing.append("Target")
    if not _select_present(page_properties.get("Time Horizon")):
        missing.append("Time Horizon")
    for gate in REQUIRED_BOOLEAN_GATES:
        checkbox_set = _checkbox_true(page_properties.get(gate))
        inferred_set = bool(inferred_checks and inferred_checks.get(gate))
        if not (checkbox_set or inferred_set):
            missing.append(gate)
    return missing

File: src/workflow/gates.ts
export const GOVERNANCE_CHECKBOX_FIELDS = [
  "≥3 Options Evaluated",
  "Success Metrics Defined",
  "Leading Indicators Defined",
  "Kill Criteria Defined",
  "Option Trade-offs Explicit",
  "Risk Matrix Completed",
  "Financial Model Included",
  "Downside Modeled",
  "Compliance Reviewed",
  "Decision Memo Written",
  "Root Cause Done",
  "Assumptions Logged",
];

export const REQUIRED_BOOLEAN_GATES = [
  "Strategic Alignment Brief",
  "Problem Quantified",
  "≥3 Options Evaluated",
  "Success Metrics Defined",
  "Leading Indicators Defined",
  "Kill Criteria Defined",
];

function checkboxTrue(prop: unknown): boolean {
  if (typeof prop === "boolean") {
    return prop;
  }

  if (!prop || typeof prop !== "object" || Array.isArray(prop)) {
    return false;
  }

  const record = prop as Record<string, unknown>;
  return Boolean(record.checkbox);
}

function numberPresent(prop: unknown): boolean {
  if (typeof prop === "number" && Number.isFinite(prop)) {
    return true;
  }

  if (!prop || typeof prop !== "object" || Array.isArray(prop)) {
    return false;
  }

  const record = prop as Record<string, unknown>;
  if (typeof record.number === "number" && Number.isFinite(record.number)) {
    return true;
  }

  return false;
}

function selectPresent(prop: unknown): boolean {
  if (typeof prop === "string") {
    return prop.trim().length > 0;
  }

  if (!prop || typeof prop !== "object" || Array.isArray(prop)) {
    return false;
  }

  const record = prop as Record<string, unknown>;
  const select = record.select as Record<string, unknown> | null | undefined;
  if (typeof select?.name === "string" && select.name.trim().length > 0) {
    return true;
  }

  const status = record.status as Record<string, unknown> | null | undefined;
  if (typeof status?.name === "string" && status.name.trim().length > 0) {
    return true;
  }

  if (typeof record.name === "string" && record.name.trim().length > 0) {
    return true;
  }

  return false;
}

export function inferGovernanceChecksFromText(bodyText: string): Record<string, boolean> {
  const text = bodyText.toLowerCase();

  const hasAny = (...phrases: string[]): boolean => phrases.some((phrase) => text.includes(phrase));
  const explicitlyNo = (gateName: string): boolean => text.includes(`${gateName.toLowerCase()}: no`);

  const optionMatches = text.match(/\boption\s+[a-z0-9]+\b/g) ?? [];
  const numericMatches = text.match(/\b\d[\d,.%]*\b/g) ?? [];

  return {
    "Strategic Alignment Brief": !explicitlyNo("Strategic Alignment Brief") && hasAny("strategic context", "strategic alignment", "objective supported"),
    "Problem Quantified":
      !explicitlyNo("Problem Quantified") &&
      hasAny("problem framing", "quantified impact", "problem statement") &&
      numericMatches.length >= 3,
    "≥3 Options Evaluated":
      !explicitlyNo("≥3 Options Evaluated") &&
      hasAny("options evaluated", "chosen option") &&
      new Set(optionMatches).size >= 3,
    "Success Metrics Defined": !explicitlyNo("Success Metrics Defined") && hasAny("success metrics", "primary metric", "kpi impact"),
    "Leading Indicators Defined": !explicitlyNo("Leading Indicators Defined") && hasAny("leading indicators"),
    "Kill Criteria Defined": !explicitlyNo("Kill Criteria Defined") && hasAny("kill criteria", "we will stop or pivot"),
    "Option Trade-offs Explicit": !explicitlyNo("Option Trade-offs Explicit") && hasAny("trade-offs", "trade offs"),
    "Risk Matrix Completed":
      !explicitlyNo("Risk Matrix Completed") && hasAny("risk matrix") && hasAny("mitigation", "probability", "impact"),
    "Financial Model Included":
      !explicitlyNo("Financial Model Included") && hasAny("financial model", "payback period", "revenue impact", "cost impact"),
    "Downside Modeled": !explicitlyNo("Downside Modeled") && hasAny("downside", "risk-adjusted", "sensitivity"),
    "Compliance Reviewed":
      !explicitlyNo("Compliance Reviewed") && hasAny("compliance review", "compliance reviewed", "legal review", "regulatory review"),
    "Decision Memo Written": !explicitlyNo("Decision Memo Written") && hasAny("executive summary", "final decision"),
    "Root Cause Done": !explicitlyNo("Root Cause Done") && hasAny("root cause"),
    "Assumptions Logged": !explicitlyNo("Assumptions Logged") && hasAny("assumptions", "confidence level"),
  };
}

export function evaluateRequiredGates(
  pageProperties: Record<string, unknown>,
  inferredChecks: Record<string, boolean> | null = null,
): string[] {
  const missing: string[] = [];

  if (!numberPresent(pageProperties.Baseline as Record<string, unknown> | null | undefined)) {
    missing.push("Baseline");
  }

  if (!numberPresent(pageProperties.Target as Record<string, unknown> | null | undefined)) {
    missing.push("Target");
  }

  if (!selectPresent(pageProperties["Time Horizon"] as Record<string, unknown> | null | undefined)) {
    missing.push("Time Horizon");
  }

  for (const gate of REQUIRED_BOOLEAN_GATES) {
    const checkboxSet = checkboxTrue(pageProperties[gate] as Record<string, unknown> | null | undefined);
    const inferredSet = Boolean(inferredChecks?.[gate]);
    if (!checkboxSet && !inferredSet) {
      missing.push(gate);
    }
  }

  return missing;
}
